File: 200/211 - Add and Search Word - Data structure design.java
public class WordDictionary {
    
    private class TrieNode {
        private boolean isWord;
        private TrieNode[] children;
        
        public TrieNode() {
            this.children = new TrieNode[26];
            this.isWord = false;
        }
        
        public void setWord() {
            this.isWord = true;
        }
        
        public boolean isWord() {
            return this.isWord;
        }
        
        public void setChild(int index) {
            this.children[index] = new TrieNode();
        }
        
        public TrieNode getChild(int index) {
            return this.children[index];
        }
    }

    private TrieNode root;
    
    /** Initialize your data structure here. */
    public WordDictionary() {
        this.root = new TrieNode();
    }
    
    /** Adds a word into the data structure. */
    public void addWord(String word) {
        TrieNode p = this.root;
        int i;
        
        for (char c : word.toCharArray()) {
            i = (int)(c - 'a');
            if (p.getChild(i) == null)
                p.setChild(i);
            p = p.getChild(i);
        }
        
        p.setWord();
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    public boolean search(String word) {
        return searchRecursive(word, this.root, 0);
    }
    
    private boolean searchRecursive(String word, TrieNode curr, int i) {
        char c;
        int j;
        TrieNode next;
        
        if (i == word.length())
            return curr.isWord();
        c = word.charAt(i);
        
        if (c == '.') {
            for (j = 0; j < 26; ++j) {
                next = curr.getChild(j);
                if (next != null && searchRecursive(word, next, i + 1))
                    return true;
            }
            return false;
        } else {
            next = curr.getChild((int)(c - 'a'));
            if (next == null)
                return false;
            else
                return searchRecursive(word, next, i + 1);
        }
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */
 

File: csharp/211.add-and-search-word-data-structure-design.cs
/*
 * @lc app=leetcode id=211 lang=csharp
 *
 * [211] Add and Search Word - Data structure design
 */

using System.Collections.Generic;
using System.Linq;

public class WordDictionary {

    private class TrieNode {
        public bool end { get; set; }
        public TrieNode[] children { get; }

        public TrieNode() {
            end = false;
            children = new TrieNode[26];
        }
    }

    private TrieNode root;

    /** Initialize your data structure here. */
    public WordDictionary() {
        root = new TrieNode();
    }
    
    /** Adds a word into the data structure. */
    public void AddWord(string word) {
        TrieNode curr = root;
        foreach (int i in word.Select(c => (int) c - (int) 'a')) {
            if (curr.children[i] == null) {
                curr.children[i] = new TrieNode();
            }
            curr = curr.children[i];
        }
        curr.end = true;
    }

    private bool Search(TrieNode curr, int[] word, int index) {
        if (curr == null) {
            return false;
        }
        if (index == word.Length) {
            return curr.end;
        }
        return (word[index] == (int ) '.' - (int) 'a' ? curr.children : new TrieNode[]{ curr.children[word[index]] }).Any(node => Search(node, word, index + 1));
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    public bool Search(string word) {
        return Search(root, word.Select(c => (int) c - (int) 'a').ToArray(), 0);
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.AddWord(word);
 * bool param_2 = obj.Search(word);
 */

